curve/ecgfp5/scalar_field.go

```diff
@@ -0,0 +1,343 @@
+// If c == 0xFFFFFFFFFFFFFFFF, return a1.
+// c MUST be equal to 0 or 0xFFFFFFFFFFFFFFFF.
+func Select(c uint64, a0, a1 *ECgFp5Scalar) *ECgFp5Scalar {
+	return &ECgFp5Scalar{
+		a0[0] ^ (c & (a0[0] ^ a1[0])),
+		a0[1] ^ (c & (a0[1] ^ a1[1])),
+		a0[2] ^ (c & (a0[2] ^ a1[2])),
+		a0[3] ^ (c & (a0[3] ^ a1[3])),
+		a0[4] ^ (c & (a0[4] ^ a1[4])),
+	}
+}
+
+func (s ECgFp5Scalar) Add(rhs ECgFp5Scalar) ECgFp5Scalar {
+	r0 := s.AddInner(rhs)
+	r1, c := r0.SubInner(&N)
+	return *Select(c, r1, &r0)
+}
+
+func (s ECgFp5Scalar) Sub(rhs ECgFp5Scalar) ECgFp5Scalar {
+	r0, c := s.SubInner(&rhs)
+	r1 := r0.AddInner(N)
+	return *Select(c, r0, &r1)
+}
+
+func (s ECgFp5Scalar) Neg() ECgFp5Scalar {
+	return ZERO.Sub(s)
+}
+
+func (s *ECgFp5Scalar) Mul(rhs *ECgFp5Scalar) *ECgFp5Scalar {
+	res := s.MontyMul(&R2).MontyMul(rhs)
+	return res
+}
+
+func (s *ECgFp5Scalar) Square() *ECgFp5Scalar {
+	return s.Mul(s)
+}
+
+// Montgomery multiplication.
+// Returns (self*rhs)/2^320 mod n.
+// 'self' MUST be less than n (the other operand can be up to 2^320-1).
+func (s *ECgFp5Scalar) MontyMul(rhs *ECgFp5Scalar) *ECgFp5Scalar {
+	r := new(ECgFp5Scalar)
+	for i := 0; i < 5; i++ {
+		// Iteration i computes r <- (r + self*rhs_i + f*n)/2^64.
+		// Factor f is at most 2^64-1 and set so that the division
+		// is exact.
+		// On input:
+		//    r <= 2^320 - 1
+		//    self <= n - 1
+		//    rhs_i <= 2^64 - 1
+		//    f <= 2^64 - 1
+		// Therefore:
+		//    r + self*rhs_i + f*n <= 2^320-1 + (2^64 - 1) * (n - 1)
+		//                            + (2^64 - 1) * n
+		//                         < 2^384
+		// Thus, the new r fits on 320 bits.
+		m := rhs[i]
+		f := (s[0]*m + r[0]) * N0I
+
+		cc1, cc2 := uint64(0), uint64(0)
+		for j := 0; j < 5; j++ {
+			z := U128From64(s[j]).Mul64(m).Add64(r[j]).Add64(cc1)
+			cc1 = z.Hi
+			z = U128From64(f).Mul64(N[j]).Add64(z.Lo).Add64(cc2)
+			cc2 = z.Hi
+			if j > 0 {
+				r[j-1] = z.Lo
+			}
+		}
+		// No overflow here since the new r fits on 320 bits.
+		r[4] = cc1 + cc2
+	}
+	// We computed (self*rhs + ff*n) / 2^320, with:
+	//    self < n
+	//    rhs < 2^320
+	//    ff < 2^320
+	// Thus, the value we obtained is lower than 2*n. Subtracting n
+	// once (conditionally) is sufficient to achieve full reduction.
+	r2, c := r.SubInner(&N)
+	return Select(c, r2, r)
+}
+
+func (s ECgFp5Scalar) expPowerOf2(exp int) ECgFp5Scalar {
+	result := s
+	for i := 0; i < exp; i++ {
+		result = *result.Square()
+	}
+	return result
+}
+
+func FromGfp5(fp5 gFp5.Element) ECgFp5Scalar {
+	return FromNonCanonicalBigInt(BigIntFromArray([5]uint64{
+		fp5[0].Uint64(), fp5[1].Uint64(), fp5[2].Uint64(), fp5[3].Uint64(), fp5[4].Uint64(),
+	}))
+}
+
+func BigIntFromArray(arr [5]uint64) *big.Int {
+	result := new(big.Int)
+	for i := 4; i >= 0; i-- {
+		result.Lsh(result, 64)
+		result.Or(result, new(big.Int).SetUint64(arr[i]))
+	}
+	return result
+}
+
+func FromNonCanonicalBigInt(val *big.Int) ECgFp5Scalar {
+	limbs := new(big.Int).Mod(val, ORDER).Bits()
```

Comment:
I think here we can mutate the `val` and use `val.Mod(val, ORDER)` to save a new big int allocation